Teuchos:  BLAS specify the return type of Complex blas calls

## Question
@trilinos/teuchos 

I've got alot of spam (warnings) about C-linkage and return values of complex blas functions. E.g.,

```
/g/g20/jjellio/src/github/Trilinos-a/packages/teuchos/numerics/src/Teuchos_BLAS_wrappers.hpp:198:29: warning: 'zdotc_' has C-linkage specified, but returns user-defined type 'std::complex<double>' which is incompatible with C [-Wreturn-type-c-linkage]
std::complex<double> PREFIX ZDOT_F77(const int* n, const std::complex<double> x[], const int* incx, const std::complex<double> y[], const int* incy);
                            ^
/g/g20/jjellio/src/github/Trilinos-a/packages/teuchos/numerics/src/Teuchos_BLAS_wrappers.hpp:101:21: note: expanded from macro 'ZDOT_F77'
#define ZDOT_F77    F77_BLAS_MANGLE(zdotc,ZDOTC)
                    ^
/tmp/jjellio/build/cce-14.0.2_prgenv-cray_rocm-5.2.3_mpich-8.1.18_pure-cce_hip_amd-inlall-NOfunc-gfx90a_serial_atdm_complex_opt-g_libonly-FIXES/trilinos/packages/ml/src/ml_config.h:8:37: note: expanded from macro 'F77_BLAS_MANGLE'
 #define F77_BLAS_MANGLE(name,NAME) name ## _
                                    ^
<scratch space>:277:1: note: expanded from here
zdotc_
^
```


Is there a feature in Teuchos' wrappers that handles the return type better? I skimmed the code/cmake and didn't see anything. Intel discusses this, e.g., [Calling complex blas through C++](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-linux-developer-guide/top/language-specific-usage-options/mixed-language-programming-with-onemkl/call-blas-funcs-return-complex-values-in-c-code.html). And they aren't using the return value of the blas calls. I think Trilinos prefers to call the Fortran symbols directly, so it requires C-linkages for those.

@bartlettroscoe

--- a/packages/teuchos/numerics/src/Teuchos_BLAS_wrappers.hpp
+++ b/packages/teuchos/numerics/src/Teuchos_BLAS_wrappers.hpp
@@ -68,6 +68,17 @@
 #  define Teuchos_fcd const char *
 #endif
 
+// Handle Complex types (we assume C++11 at a minumum
+// Microsoft C++11 apparently does not support float/double _Complex
+
+#if ( defined(_MSC_VER) )
+#define Teuchos_Complex_double_type_name std::complex<double>
+#define Teuchos_Complex_float_type_name std::complex<float>
+#else
+#define Teuchos_Complex_double_type_name double _Complex
+#define Teuchos_Complex_float_type_name float _Complex
+#endif
+
 
 /* B) Take care of of the link name case */
 
@@ -195,7 +206,7 @@ void PREFIX ZDOT_F77(std::complex<double> *ret, const int* n, const std::complex
   // reimplementing the offending routines.
 #    endif // HAVE_COMPLEX_BLAS_PROBLEM
 #  else // no problem
-std::complex<double> PREFIX ZDOT_F77(const int* n, const std::complex<double> x[], const int* incx, const std::complex<double> y[], const int* incy);
+Teuchos_Complex_double_type_name PREFIX ZDOT_F77(const int* n, const std::complex<double> x[], const int* incx, const std::complex<double> y[], const int* incy);
 #  endif // defined(HAVE_COMPLEX_BLAS_PROBLEM)
 
 double PREFIX ZNRM2_F77(const int* n, const std::complex<double> x[], const int* incx);
@@ -255,7 +266,7 @@ float PREFIX SCNRM2_F77(const int* n, const std::complex<float> x[], const int* 
 #elif defined(HAVE_COMPLEX_BLAS_PROBLEM) && defined(HAVE_FIXABLE_COMPLEX_BLAS_PROBLEM)
 void PREFIX CDOT_F77(std::complex<float> *ret, const int* n, const std::complex<float> x[], const int* incx, const std::complex<float> y[], const int* incy);
 #elif defined(HAVE_TEUCHOS_BLASFLOAT)
-std::complex<float> PREFIX CDOT_F77(const int* n, const std::complex<float> x[], const int* incx, const std::complex<float> y[], const int* incy);
+Teuchos_Complex_float_type_name PREFIX CDOT_F77(const int* n, const std::complex<float> x[], const int* incx, const std::complex<float> y[], const int* incy);
 #else
 // the code is literally in Teuchos_BLAS.cpp
 #endif

